aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -291,6 +292,52 @@ Tensor& sum_out(Tensor& result, const Tensor& self, DimnameList dim,
   return at::sum_out(result, self, dimnames_to_positions(self, dim), keepdim, opt_dtype);
 }
 
+Tensor& nansum_out(Tensor& result, const Tensor& self, IntArrayRef dim,
+                       bool keepdim, optional<ScalarType> opt_dtype) {
+
+  auto get_computation_dtype = [](const Tensor& self) {
+    ScalarType src_type = self.scalar_type();
+    if (at::isIntegralType(src_type, /*includeBool=*/true)) {
+      return c10::typeMetaToScalarType(c10::get_default_dtype());
+    }
+    return src_type;
+  };
+
+  ScalarType scalarType =
+      opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  // numpy promotes all integral types to `Long` and preserver the floating types.
+  const auto result_type = at::isIntegralType(scalarType) ? ScalarType::Long : scalarType;
+  // use default float type for computation if `self.dtype()` is integral
+  const auto dtype = get_computation_dtype(self);
+
+  auto iter = make_reduction("nansum", result, self, dim, keepdim, dtype);
+  if (iter.numel() == 0) {
+    result = result.zero_().toType(result_type);
+  } else {
+    nansum_stub(iter.device_type(), iter);
+    result = result.toType(result_type);
+  }
+  return result;
+}
+
+Tensor nansum(const Tensor &self, c10::optional<ScalarType> dtype) {
+  return at::native::nansum(self, std::vector<int64_t>{}, false, dtype);
+}
+
+Tensor nansum(const Tensor& self, IntArrayRef dim, bool keepdim, c10::optional<ScalarType> dtype) {
+  Tensor result;
+  return at::native::nansum_out(result, self, dim, keepdim, dtype);
+}
+
+Tensor nansum(const Tensor& self, DimnameList dim, bool keepdim, c10::optional<ScalarType> dtype) {
```

Comment:
To keep this PR as simple as possible, let's not do a named variant. It can be added seperately later.

Reply:
Makes sense. Removed.